src/librustc_mir/interpret/terminator.rs

```diff
@@ -293,6 +293,31 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                     }
                 }
 
+                // If this function is a const function then as an optimisation we can query this
+                // evaluation immediately.
+                //
+                // For the moment we only do this for functions which take no arguments
+                // (or all arguments are ZSTs) so that we don't memoize too much.
+                if self.tcx.is_const_fn_raw(instance.def.def_id()) &&
+                   args.iter().all(|a| a.layout.is_zst())
+                {
+                    let gid = GlobalId { instance, promoted: None };
+                    let place = self.const_eval_raw(gid)?;
+
+                    let dest = match dest {
```

Comment:
```suggestion
                    let dest = dest.unwrap_or_else(|| throw_ub!(Unreachable));
```

Reply:
Unfortunately can't use `unwrap_or_else()` here because `throw_ub!` macro expands to something of the form `return Err(..)`.

Reply:
Actually I think `.ok_or_else(|| throw_ub!(Unreachable))?` works